[CDAP-18322] Set empty user credential in SecurityRequestContext instead of enforcing they are present in the request.

wyzhang opened the pull request:
Why:
Enforcing user credentials are always present is quite restrictive,
setting dummy user id and credential in the security request context
so they get failed when auth is enforced while they will pass when
no auth is needed (e.g. making ping request to check if services
are alive)

dli357 commented on cdap-common/src/main/java/io/cdap/cdap/common/http/AuthenticationChannelHandler.java:
```diff
@@ -86,23 +93,17 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-          throw new IllegalArgumentException("Missing user credential for HTTP content request");
-        }
-      }
+      //
```
nit: empty comment

wyzhang replied:
Removed

dli357 commented on cdap-common/src/main/java/io/cdap/cdap/common/http/AuthenticationChannelHandler.java:
```diff
@@ -86,23 +93,17 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
+      LOG.trace("Got user ID '{}', user IP '{}', and authorization header length '{}'",
+                userID == null ? "NULL" : userID,
+                userIP == null ? "NULL" : userIP,
+                authHeader == null ? "NULL" : String.valueOf(authHeader.length()));
```
We should also log the source IP of the channel request using `ctx.channel().remoteAddress().toString()` as was previously logged in line 63 to allow us to better debug where a request came from when it is propagated through multiple services.

wyzhang replied:
Done. Added.

dli357 reviewed (approved):
I think we should always require a system ID and credential even when pinging to check if system services are alive, but that can be done in a later iteration. LGTM pending resolution of the other two comments, thanks!

chtyim commented on cdap-common/src/main/java/io/cdap/cdap/common/http/AuthenticationChannelHandler.java:
```diff
@@ -86,27 +91,21 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-
-      SecurityRequestContext.setUserId(currentUserId);
+      LOG.trace("Got user ID '{}' user IP '{}' from IP '{}' and authorization header length '{}'",
+                userID == null ? "NULL" : userID,
```
I think you don't need the `: ?`. Having a `null` value would resulted in a `null` string in the log text.

wyzhang replied:
Fixed

chtyim commented on cdap-common/src/main/java/io/cdap/cdap/common/http/AuthenticationChannelHandler.java:
```diff
@@ -86,27 +91,21 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
+      LOG.trace("Got user ID '{}' user IP '{}' from IP '{}' and authorization header length '{}'",
+                userID == null ? "NULL" : userID,
+                userIP == null ? "NULL" : userIP,
+                ctx.channel().remoteAddress().toString(),
```
No need to call `.toString()`. The slf4j `{}` placeholder will do that when forming the final log message. This is the way to avoid heavy `.toString()` if the given log level is not enabled.

wyzhang replied:
Fixed.

chtyim commented on cdap-common/src/main/java/io/cdap/cdap/common/http/AuthenticationChannelHandler.java:
```diff
@@ -123,4 +122,16 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
+  private void resetCurrentUserInfo() {
+    if (internalAuthEnabled) {
+      currentUserID = EMPTY_USER_ID;
+      currentUserCredential = new Credential(EMPTY_USER_CREDENTIAL, Credential.CredentialType.INTERNAL);
```
Seems like the `Credential` object can be a constant too?

wyzhang replied:
Fixed.

wyzhang committed:
[CDAP-18322] Set empty user credential in SecurityRequestContext instead of
             enforcing they are present in the request when internal auth is
             enabled.

Why:
Enforcing user credentials are always present is quite restrictive,
setting dummy user id and credential in the security request context
so they get failed when auth is enforced while they will pass when
no auth is needed (e.g. making ping request to check if services
are alive)
